uvicorn/protocols/websockets/websockets_impl.py

```diff
@@ -145,6 +144,10 @@ def connection_lost(self, exc: Optional[Exception]) -> None:
 
     def shutdown(self) -> None:
         self.ws_server.closing = True
+        if self.handshake_completed_event.is_set():
+            self.fail_connection(1012)
+        else:
+            self.send_500_response()
```

Comment:
This is the change that motivated this PR.

We were only sending 1006 to the client, even when the handshake was not completed.

Reply:
Why `self.fail_connection(1012)` instead of just calling `self.close(1012)`?

I've been look at `websockets` [server.close()](https://github.com/aaugustin/websockets/blob/e4fcab16a70344e356268e50dc0c0cf541920c5d/src/websockets/legacy/server.py#L780-L783) and [procotol.close()](https://github.com/aaugustin/websockets/blob/e4fcab16a70344e356268e50dc0c0cf541920c5d/src/websockets/legacy/protocol.py#L741-L796) to see why uvicorn isn't calling `close(), but haven't gotten terribly far yet.

Reply:
Because `close` is async.

Reply:
Ahh, missed the lack of async here.

Looking deeper at the difference, it _seems_ like the primary difference is `fail_connection` will proactively cancel the data transfer task but `close` will simply send the close frame and then wait.

Is there any consequence to that distinction here? My very, very basic testing seems to suggest with uvicorn tasks are given a chance to shutdown cleanly, but I'm still testing.

Reply:
Yeah, this code still waits for it to finish. :+1: